compiler/rustc_trait_selection/src/solve/trait_goals.rs

```diff
@@ -238,14 +242,25 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
         ecx: &mut EvalCtxt<'_, 'tcx>,
         goal: Goal<'tcx, Self>,
     ) -> QueryResult<'tcx> {
-        if goal.predicate.polarity != ty::ImplPolarity::Positive {
-            return Err(NoSolution);
-        }
-
-        if let ty::FnPtr(..) = goal.predicate.self_ty().kind() {
-            ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
-        } else {
-            Err(NoSolution)
+        let self_ty = goal.predicate.self_ty();
+        match goal.predicate.polarity {
+            ty::ImplPolarity::Positive => {
+                if self_ty.is_fn_ptr() {
+                    ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+                } else {
+                    Err(NoSolution)
+                }
+            }
+            ty::ImplPolarity::Negative => {
+                // If a type is rigid and not a fn ptr, then we know for certain
+                // that it does *not* implement `FnPtr`.
+                if !self_ty.is_fn_ptr() && self_ty.is_known_rigid() {
+                    ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
```

Comment:
should we return `Certainty::Ambiguous` if the `self_ty` is not rigid? i can't think of how this would cause issues, but it does feel safer to me

Reply:
Well, we don't assemble an ambiguous candidate for non-rigid self types when the impl polarity is positive, so I wouldn't see why we need to do this for negative impls. But I can add this with a FIXME saying it's unnecessary and probably should be removed.

Reply:
it feels like it only matters for `<T as Trait>::Assoc: !FnPtr` which I guess doesn't exist in intercrate mode as there's an empty env?

I guess you're right and we don't need it